### ISSUE TITLE
Implement renaming symbols via LSP and tsserver

### ISSUE THREAD

**Issue by w0rp** [dense-analysis/ale#1519]
LSP and tsserver support operations for renaming things, and ALE should support these operations.

**Comment by mwilliammyers** [comment]
I am interested in tackling this...

**Comment by wavded** [comment]
This would be super handy to have. Any plans still @w0rp ? I would prefer to just use ale than have another lang server plugin that supports this.

**Comment by w0rp** [comment]
The plan is to just eventually implement it somehow.

**Comment by eugene-bright** [comment]
LanguageClient-neovim implements this feature. It's a subject to some 'code sharing'.
https://github.com/autozimu/LanguageClient-neovim/issues/35#issuecomment-288731403

**Comment by w0rp** [comment]
It sure does. It's not ALE. I'll implement this eventually, or someone else can try.

**Comment by eugene-bright** [comment]
May we get the feature by low price by passing LSP connection to LanguageClient-neovim functions? Raw ugly but possible working idea.

**Comment by mwilliammyers** [comment]
I have an old [commit](https://github.com/mwilliammyers/ale/commit/5eb8d3) that adds support for the lsp and tsserver renaming protocols. 

I haven't added support for actually performing the rename in vim and probably won't have time to work on it for a while, but I have rebased my branch onto upstream so I could open a work in progress PR or someone is welcome to use [my commit](https://github.com/mwilliammyers/ale/commit/5eb8d3) as a basis...

**Comment by w0rp** [comment]
Thank you for sharing that. Someone might pick it up and finish the job. :+1:

**Comment by Risto-Stevcev** [comment]
I tried getting this to work a while back but it doesn't seem to rename anything:
https://github.com/Risto-Stevcev/ale/commit/a94bf9b095d8a542f17c9ceeece311d2ab283077

This is such a critical IDE feature, it would be nice to get this in

**Comment by jeremija** [comment]
I made rename work for TypeScript by starting off of @mwilliammyers's commit (thanks!). It's on my fork on the [feature/renaming-symbols](https://github.com/jeremija/ale/compare/feature/renaming-symbols) branch. Here's the link to the main [commit](https://github.com/jeremija/ale/commit/2e6e77a87b2931a4966112fa531177360cc5d6b4). I also merged `upstream/master` into it to keep it up to date.

Since I changed the message structure in `ale#rename#HandleTSServerResponse` (that eventually gets passed on to `s:ApplyRenameEdits`, it also needs to be changed in `ale#rename#HandleLSPResponse`.

Also, the way the [buffer contents get replaced](https://github.com/jeremija/ale/commit/2e6e77a87b2931a4966112fa531177360cc5d6b4#diff-f3ec3b71f8497b3a282a0d2845147db2R32) seems hackish as hell, so I'd love to hear about a better way to do it :)
